extension/src/utils/aria.ts

```diff
@@ -134,3 +141,36 @@ export function getElementRoles(element: Element): Set<ElementRole> {
 
   return new Set(matchedRoles?.flatMap((role) => role.roles))
 }
+
+function getLabelTexts(element: Element): string[] {
+  if (
+    element instanceof HTMLInputElement === false &&
+    element instanceof HTMLTextAreaElement === false &&
+    element instanceof HTMLSelectElement === false
+  ) {
+    return []
+  }
+
+  if (element.labels === null) {
+    return []
+  }
+
+  return Array.from(element.labels)
+    .map((label) => label.textContent)
+    .filter((label) => label !== null)
+}
+
+export function getAriaDetails(element: Element): AriaDetails {
+  const roles = [...getElementRoles(element)]
+    .filter((r) => !ABSTRACT_ROLES.has(r.role))
```

Comment:
Some roles in ARIA are abstract, e.g. `input` which is implemented by `checkbox`, `radio` and `textbox`, which means that they aren't specific enough to be used for querying the DOM.